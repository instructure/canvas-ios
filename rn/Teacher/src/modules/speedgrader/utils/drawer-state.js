//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import { Animated, Keyboard } from 'react-native'

export type DrawerPosition = 0 | 1 | 2
export type DrawerObserver = {
  snapTo: (position: DrawerPosition, animated?: boolean) => void,
  onDragBegan?: () => void,
}

export const HANDLE_PADDING_BOTTOM: number = 52
export const HANDLE_INVISIBLE_TOP_PADDING: number = 16
export const HANDLE_VISIBLE_PADDING: number = 6
export const HANDLE_PADDING_TOP: number = HANDLE_INVISIBLE_TOP_PADDING + HANDLE_VISIBLE_PADDING
export const HANDLE_BAR_HEIGHT: number = 4
export const HANDLE_HEIGHT: number = HANDLE_PADDING_TOP + HANDLE_PADDING_BOTTOM + HANDLE_BAR_HEIGHT
// space the drawer will not enter when fully open
const DRAWER_MIN_TOP_PADDING = 60 - HANDLE_INVISIBLE_TOP_PADDING

export default class DrawerState {
  currentSnap: DrawerPosition
  deltaY: Animated.Value
  drawers: Array<DrawerObserver>
  commentProgress: { [string]: Animated.Value }

  drawerHeight (position: DrawerPosition, screenHeight: number): number {
    return this.drawerSnapPoint(position, screenHeight) + HANDLE_HEIGHT - HANDLE_INVISIBLE_TOP_PADDING
  }

  drawerSnapPoint (position: DrawerPosition, screenHeight: number): number {
    const minHeight = Math.max(screenHeight, 140) // for tests
    switch (position) {
      case 1: return minHeight * 0.5 - HANDLE_PADDING_BOTTOM
      case 2: return minHeight - HANDLE_HEIGHT - DRAWER_MIN_TOP_PADDING
      default: return 0
    }
  }

  constructor () {
    this.currentSnap = 0
    this.deltaY = new Animated.Value(0)
    this.drawers = []
    this.commentProgress = {}
  }

  registerDrawer (drawer: DrawerObserver) {
    drawer.snapTo(this.currentSnap, false)
    this.drawers.push(drawer)
  }

  unregisterDrawer (drawer: DrawerObserver) {
    this.drawers = this.drawers.filter(registered => registered !== drawer)
  }

  dragBegan = () => {
    this.drawers.forEach(drawer => drawer.onDragBegan && drawer.onDragBegan())
  }

  drawerDidSnap = (drawer: DrawerObserver, snap: DrawerPosition) => {
    this.snapTo(snap, false, drawer)
    if (snap !== 2) {
      Keyboard.dismiss()
    }
  }

  snapTo = (snap: DrawerPosition, animated: boolean = true, initiatingDrawer?: DrawerObserver) => {
    if (snap !== this.currentSnap) {
      this.currentSnap = snap
      this.drawers.forEach(drawer => drawer !== initiatingDrawer && drawer.snapTo(snap, animated))
      if (!animated && snap === 0) {
        this.deltaY.setValue(0)
      }
    }
  }

  registerCommentProgress = (userID: string, value: Animated.Value) => {
    this.commentProgress[userID] = value
  }

  unregisterCommentProgress = (userID: string) => {
    delete this.commentProgress[userID]
  }
}
