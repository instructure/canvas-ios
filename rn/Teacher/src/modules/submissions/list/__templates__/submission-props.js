//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import template, { type Template } from '../../../../utils/template'
import type { SubmissionDataProps } from '../submission-prop-types'

const templates = {
  ...require('../../../../__templates__/submissions'),
}

const submissionWithHistory = templates.submissionHistory()

// $FlowFixMe
export const submissionProps: Template<SubmissionDataProps> = template({
  grade: 'B-',
  name: submissionWithHistory.user.name,
  status: 'submitted',
  userID: submissionWithHistory.user.id,
  avatarURL: submissionWithHistory.user.avatar_url,
  sectionID: '1',
})
