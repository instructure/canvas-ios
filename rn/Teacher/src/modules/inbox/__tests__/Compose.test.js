//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* eslint-disable flowtype/require-valid-file-annotation */

import React from 'react'
import { NativeModules } from 'react-native'
import { shallow } from 'enzyme'
import { Compose, mapStateToProps } from '../Compose'
import api from '../../../canvas-api'
import { apiResponse } from '../../../canvas-api/utils/testHelpers'

import * as template from '../../../__templates__'

let defaultProps = {
  navigator: template.navigator({
    dismiss: jest.fn(),
  }),
  refreshInboxSent: jest.fn(),
}

jest
  .mock('react-native/Libraries/LayoutAnimation/LayoutAnimation', () => ({
    configureNext: jest.fn(),
    easeInEaseOut: jest.fn(),
    Types: {
      easeInEaseOut: jest.fn(),
      spring: jest.fn(),
    },
    Properties: {
      opacity: 1,
    },
  }))
  .mock('react-native/Libraries/Components/Touchable/TouchableOpacity', () => 'TouchableOpacity')
  .mock('react-native/Libraries/Alert/Alert', () => ({ alert: jest.fn() }))
  .mock('../../../canvas-api')
  .mock('../../../routing/Screen')

describe('Compose', () => {
  beforeEach(() => jest.clearAllMocks())

  it('renders', () => {
    let tree = shallow(<Compose {...defaultProps} />)
    expect(tree).toMatchSnapshot()
  })

  it('allows for a navbar title to be passed in as a prop', () => {
    let tree = shallow(<Compose {...defaultProps} navBarTitle='title' />)
    expect(tree.find('Screen').prop('title')).toEqual('title')
  })

  it('doesnt show the course select when told not to', () => {
    let tree = shallow(<Compose {...defaultProps} showCourseSelect={false} />)
    expect(tree.find('[testID="compose.course-select"]').exists()).toBe(false)
  })

  it('renders with passed in recipients', () => {
    const u1 = template.addressBookResult({ id: '1' })
    const u2 = template.addressBookResult({ id: '2' })
    let tree = shallow(<Compose {...defaultProps} recipients={[u1, u2]} />)
    expect(tree).toMatchSnapshot()
  })

  it('doesnt show the To placeholder when there are recipients', () => {
    const recipient = template.addressBookResult({ id: '1' })
    let tree = shallow(<Compose {...defaultProps} recipients={[recipient]} />)
    expect(tree.find('[testID="compose.recipients-placeholder"]').exists()).toBe(false)
  })

  it('renders after picking a course', () => {
    let course = template.course()
    let onSelect = jest.fn()
    const show = jest.fn((path, options, passthrough) => {
      onSelect = passthrough.onSelect
    })

    const navigator = template.navigator({ show, dismiss: jest.fn() })
    let tree = shallow(<Compose {...defaultProps} navigator={navigator} />)

    tree.find('[testID="compose.course-select"]').simulate('Press')
    onSelect(course)
    expect(navigator.pop).toHaveBeenCalled()
    expect(tree).toMatchSnapshot()
  })

  it('sets all the data and then send the message', () => {
    let course = template.course()
    const recipient = template.addressBookResult()
    let component = shallow(<Compose {...defaultProps} />)
    const instance = component.instance()
    instance._bodyChanged('body of the message')
    instance._subjectChanged('subject of the message')
    instance.setStateAndUpdate({ contextName: course.name, contextCode: `course_${course.id}` })
    instance.setStateAndUpdate({ recipients: [recipient] })
    expect(instance.state.sendDisabled).toEqual(false)
  })

  it('sets recipients to teacher after selecting course if instructorQuestion', async () => {
    const course = template.course({ id: '1' })
    const props = {
      ...defaultProps,
      instructorQuestion: true,
      navigator: template.navigator({
        show: jest.fn((path, options, passthrough) => {
          passthrough.onSelect(course)
        }),
      }),
    }
    const view = shallow(<Compose {...props} />)
    const courseSelect = view.find('[testID="compose.course-select"]')
    courseSelect.simulate('Press')
    await view.update()
    const token = view.find('AddressBookToken')
    expect(token.length).toEqual(1)
    expect(token.prop('item').id).toEqual('course_1_teachers')
    expect(token.prop('item').name).toEqual('Teachers')
    expect(token.prop('canDelete')).toEqual(false)
  })

  it('can delete receipients if !instructorQuestion', () => {
    const props = {
      ...defaultProps,
      contextCode: 'course_1',
      instructorQuestion: false,
      recipients: [template.addressBookResult()],
    }
    const view = shallow(<Compose {...props} />)
    const token = view.find('AddressBookToken')
    expect(token.length).toEqual(1)
    expect(token.prop('canDelete')).toEqual(true)
  })

  it('hides option to send individual messages when instructorQuestion', () => {
    const props = {
      ...defaultProps,
      instructorQuestion: true,
    }
    const view = shallow(<Compose {...props} />)
    const option = view.find('[identifier="compose-message.send-all-toggle"]')
    expect(option.exists()).toEqual(false)
  })

  it('shows option to send individual messages if !instructorQuestion', () => {
    const props = {
      ...defaultProps,
      instructorQuestion: false,
    }
    const view = shallow(<Compose {...props} />)
    const option = view.find('[identifier="compose-message.send-all-toggle"]')
    expect(option.exists()).toEqual(true)
  })

  it('gets and sets recipients', () => {
    const recipient = template.addressBookResult()
    let onSelect = jest.fn()
    let onCancel = jest.fn()
    const show = jest.fn((path, options, passthrough) => {
      onSelect = passthrough.onSelect
      onCancel = passthrough.onCancel
    })

    const navigator = template.navigator({ show, dismiss: jest.fn() })
    let component = shallow(<Compose {...defaultProps} navigator={navigator} />)

    component.instance().selectCourse()
    onSelect(template.course())
    component.instance()._openAddressBook()
    onSelect([recipient])
    expect(navigator.dismiss).toHaveBeenCalled()
    expect(component.instance().state.recipients).toEqual([recipient])
    onCancel()
    expect(navigator.dismiss).toHaveBeenCalled()
  })

  it('dismisses the modal when cancel is pressed', () => {
    let tree = shallow(<Compose {...defaultProps} />)
    tree.find('Screen').prop('leftBarButtons')[0].action()
    expect(defaultProps.navigator.dismiss).toHaveBeenCalled()
  })

  it('toggles the send to all', () => {
    let tree = shallow(<Compose {...defaultProps} />)
    tree.find('[identifier="compose-message.send-all-toggle"]').simulate('ValueChange', true)
    expect(tree.state('sendToAll')).toBeTruthy()
  })

  it('hides send to all for replies', () => {
    let tree = shallow(<Compose {...defaultProps} conversationID='2' />)
    expect(tree.find('[identifier="compose-message.send-all-toggle"]').exists()).toBe(false)
  })

  it('deletes a recipient from state', () => {
    let tree = shallow(<Compose {...defaultProps} />)
    tree.setState({
      recipients: [
        {
          id: '1',
          name: 'Donald Trump',
        },
      ],
    })
    tree.instance()._deleteRecipient('1')
    expect(tree.state('recipients').length).toBe(0)
  })

  it('creates conversation on send', async () => {
    const u1 = template.addressBookResult({ id: '1' })
    const props = {
      ...defaultProps,
      contextCode: 'course_1',
      recipients: [u1],
      subject: 'new conversation subject',
      onlySendIndividualMessages: false,
    }
    const dismiss = Promise.resolve()
    props.navigator.dismissAllModals = jest.fn(() => dismiss)

    let response = apiResponse(template.conversation())()
    api.createConversation.mockReturnValueOnce(response)

    const tree = shallow(<Compose {...props} />)
    tree.find('[testID="compose-message.body-text-input"]').simulate('ChangeText', 'new conversation')
    tree.find('Screen').prop('rightBarButtons')[0].action()
    expect(api.createConversation).toHaveBeenCalledWith({
      recipients: ['1'],
      body: 'new conversation',
      subject: 'new conversation subject',
      group_conversation: true,
      attachment_ids: [],
      context_code: 'course_1',
    })
    await response
    expect(props.navigator.dismissAllModals).toHaveBeenCalled()
    await dismiss
    expect(NativeModules.AppStoreReview.handleSuccessfulSubmit).toHaveBeenCalled()
  })

  it('adds message on send', () => {
    const u1 = template.addressBookResult({ id: '1' })
    const props = {
      ...defaultProps,
      recipients: [u1],
      subject: 'new conversation subject',
      onlySendIndividualMessages: true,
      conversationID: '1',
      includedMessages: [template.conversationMessage({ id: '1' }), template.conversationMessage({ id: '2' })],
      contextCode: 'course_1',
    }

    let response = apiResponse(template.conversation(props.includedMessages[0]))
    api.addMessage.mockReturnValueOnce(response())

    const tree = shallow(<Compose {...props} />)
    tree.find('[testID="compose-message.body-text-input"]').simulate('ChangeText', 'new conversation')
    tree.find('Screen').prop('rightBarButtons')[0].action()
    expect(api.addMessage).toHaveBeenCalledWith('1', {
      recipients: ['1'],
      body: 'new conversation',
      subject: 'new conversation subject',
      group_conversation: true,
      bulk_message: 1,
      included_messages: ['1', '2'],
      attachment_ids: [],
      context_code: 'course_1',
    })
  })

  it('adds message with attachments on send', () => {
    const u1 = template.addressBookResult({ id: '1' })
    const props = {
      ...defaultProps,
      recipients: [u1],
      subject: 'new conversation subject',
      onlySendIndividualMessages: false,
      contextCode: 'course_1',
      navigator: template.navigator({
        show: jest.fn((route, options, props) => {
          props.onComplete([template.attachment({ id: '234' })])
        }),
      }),
    }

    let response = apiResponse(template.conversation())
    api.createConversation.mockReturnValueOnce(response())

    const tree = shallow(<Compose {...props} />)
    tree.find('Screen').prop('rightBarButtons')[1].action()
    tree.find('Screen').prop('rightBarButtons')[0].action()
    expect(api.createConversation).toHaveBeenCalledWith({
      recipients: ['1'],
      body: '',
      subject: 'new conversation subject',
      group_conversation: true,
      attachment_ids: ['234'],
      context_code: 'course_1',
    })
  })

  it('refreshes conversation on unmount if adding reply', () => {
    const props = {
      ...defaultProps,
      refreshConversationDetails: jest.fn(),
      conversationID: '1',
    }
    const tree = shallow(<Compose {...props} />)
    tree.unmount()
    expect(props.refreshConversationDetails).toHaveBeenCalledWith('1')
  })

  it('does not refresh conversation on unmount', () => {
    const props = {
      ...defaultProps,
      refreshConversationDetails: jest.fn(),
    }
    const tree = shallow(<Compose {...props} />)
    tree.unmount()
    expect(props.refreshConversationDetails).not.toHaveBeenCalled()
  })

  it('renders reply', () => {
    const props = {
      ...defaultProps,
      contextName: 'Course 1',
      contextCode: 'course_1',
      canSelectCourse: false,
      canEditSubject: false,
    }
    expect(shallow(<Compose {...props} />)).toMatchSnapshot()
  })

  it('renders the forwarded message body', () => {
    let includedMessages = [template.conversationMessage({ body: 'yo' })]
    let tree = shallow(<Compose {...defaultProps} includedMessages={includedMessages} />)
    let forwardedMessage = tree.find('[testID="compose.forwarded-message"]')
    expect(forwardedMessage.exists()).toBe(true)
    expect(forwardedMessage.find('Text').last().children().text()).toEqual(includedMessages[0].body)
  })

  it('mapStateToProps', () => {
    const result = mapStateToProps({})
    expect(result).toEqual({})
  })
})
