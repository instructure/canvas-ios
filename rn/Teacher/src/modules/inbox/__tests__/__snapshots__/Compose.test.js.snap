// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compose renders 1`] = `
<Screen
  drawUnderNavBar={true}
  leftBarButtons={
    Array [
      Object {
        "action": [Function],
        "testID": "compose-message.cancel",
        "title": "Cancel",
      },
    ]
  }
  rightBarButtons={
    Array [
      Object {
        "action": [Function],
        "disabled": true,
        "style": "done",
        "testID": "compose-message.send",
        "title": "Send",
      },
      Object {
        "accessibilityLabel": "Edit attachments (0)",
        "action": [Function],
        "badge": false,
        "image": Object {
          "testUri": "../../../src/images/paperclip-icon.png",
        },
        "testID": "compose-message.attach",
      },
    ]
  }
  showDismissButton={false}
  title="New Message"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <ModalOverlay
      activityIndicatorColor="#fff"
      animationType="fade"
      backgroundColor="rgba(0,0,0,0.8)"
      height={170}
      showActivityIndicator={true}
      text="Sending..."
      transparent={true}
      visible={false}
      width={170}
    />
    <KeyboardAwareScrollView
      contentContainerStyle={
        Object {
          "flexGrow": 1,
          "paddingBottom": 16,
        }
      }
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <ForwardRef
        onPress={[Function]}
        style={
          Object {
            "borderBottomColor": "#C7CDD1",
            "borderBottomWidth": 0.5,
            "paddingHorizontal": 16,
          }
        }
        testID="compose.course-select"
        underlayColor="#ffffff00"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 54,
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#8B969E",
                  "fontSize": 16,
                  "lineHeight": 19,
                },
                undefined,
              ]
            }
          >
            Select a Course
          </Text>
          <DisclosureIndicator />
        </View>
      </ForwardRef>
      <View
        style={
          Object {
            "borderBottomColor": "#C7CDD1",
            "borderBottomWidth": 0.5,
            "paddingHorizontal": 16,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          onChangeText={[Function]}
          placeholder="Subject"
          placeholderTextColor="#8B969E"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "color": "#2D3B45",
                "fontSize": 16,
                "height": 54,
                "lineHeight": 19,
              },
              Object {
                "color": "#8B969E",
                "fontSize": 16,
                "lineHeight": 19,
              },
              Object {
                "color": "#2D3B45",
              },
            ]
          }
          testID="compose-message.subject-text-input"
          underlineColorAndroid="transparent"
          value={null}
        />
      </View>
      <RowWithSwitch
        border="bottom"
        identifier="compose-message.send-all-toggle"
        onValueChange={[Function]}
        title="Send individual message to each recipient"
        value={false}
      />
      <TextInput
        allowFontScaling={true}
        multiline={true}
        onChangeText={[Function]}
        placeholder="Compose Message"
        placeholderTextColor="#8B969E"
        rejectResponderTermination={true}
        scrollEnabled={false}
        style={
          Object {
            "color": "#2D3B45",
            "fontSize": 16,
            "lineHeight": 19,
            "paddingBottom": 8,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 10,
          }
        }
        testID="compose-message.body-text-input"
        underlineColorAndroid="transparent"
      />
    </KeyboardAwareScrollView>
  </View>
</Screen>
`;

exports[`Compose renders after picking a course 1`] = `
<Screen
  drawUnderNavBar={true}
  leftBarButtons={
    Array [
      Object {
        "action": [Function],
        "testID": "compose-message.cancel",
        "title": "Cancel",
      },
    ]
  }
  rightBarButtons={
    Array [
      Object {
        "action": [Function],
        "disabled": true,
        "style": "done",
        "testID": "compose-message.send",
        "title": "Send",
      },
      Object {
        "accessibilityLabel": "Edit attachments (0)",
        "action": [Function],
        "badge": false,
        "image": Object {
          "testUri": "../../../src/images/paperclip-icon.png",
        },
        "testID": "compose-message.attach",
      },
    ]
  }
  showDismissButton={false}
  title="New Message"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <ModalOverlay
      activityIndicatorColor="#fff"
      animationType="fade"
      backgroundColor="rgba(0,0,0,0.8)"
      height={170}
      showActivityIndicator={true}
      text="Sending..."
      transparent={true}
      visible={false}
      width={170}
    />
    <KeyboardAwareScrollView
      contentContainerStyle={
        Object {
          "flexGrow": 1,
          "paddingBottom": 16,
        }
      }
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <ForwardRef
        onPress={[Function]}
        style={
          Object {
            "borderBottomColor": "#C7CDD1",
            "borderBottomWidth": 0.5,
            "paddingHorizontal": 16,
          }
        }
        testID="compose.course-select"
        underlayColor="#ffffff00"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 54,
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#8B969E",
                  "fontSize": 16,
                  "lineHeight": 19,
                },
                Object {
                  "color": "#2D3B45",
                },
              ]
            }
          >
            Learn React Native
          </Text>
          <DisclosureIndicator />
        </View>
      </ForwardRef>
      <View
        style={
          Array [
            Object {
              "borderBottomColor": "#C7CDD1",
              "borderBottomWidth": 0.5,
              "paddingHorizontal": 16,
            },
            Object {
              "alignItems": "flex-start",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "minHeight": 54,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "height": 54,
              "justifyContent": "center",
              "padding": 6,
              "paddingLeft": 0,
            }
          }
          testID="compose.recipients-placeholder"
        >
          <Text
            style={
              Object {
                "color": "#8B969E",
                "fontSize": 16,
                "lineHeight": 19,
              }
            }
          >
            To
          </Text>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "flexWrap": "wrap",
              "paddingVertical": 8,
            }
          }
        />
        <TouchableOpacity
          accessibilityLabel="Add recipient"
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          onPress={[Function]}
          style={
            Object {
              "height": 54,
              "justifyContent": "center",
            }
          }
          testID="compose.add-recipient"
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/images/Add.png",
              }
            }
            style={
              Object {
                "tintColor": undefined,
              }
            }
          />
        </TouchableOpacity>
      </View>
      <View
        style={
          Object {
            "borderBottomColor": "#C7CDD1",
            "borderBottomWidth": 0.5,
            "paddingHorizontal": 16,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          onChangeText={[Function]}
          placeholder="Subject"
          placeholderTextColor="#8B969E"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "color": "#2D3B45",
                "fontSize": 16,
                "height": 54,
                "lineHeight": 19,
              },
              Object {
                "color": "#8B969E",
                "fontSize": 16,
                "lineHeight": 19,
              },
              Object {
                "color": "#2D3B45",
              },
            ]
          }
          testID="compose-message.subject-text-input"
          underlineColorAndroid="transparent"
          value={null}
        />
      </View>
      <RowWithSwitch
        border="bottom"
        identifier="compose-message.send-all-toggle"
        onValueChange={[Function]}
        title="Send individual message to each recipient"
        value={false}
      />
      <TextInput
        allowFontScaling={true}
        multiline={true}
        onChangeText={[Function]}
        placeholder="Compose Message"
        placeholderTextColor="#8B969E"
        rejectResponderTermination={true}
        scrollEnabled={false}
        style={
          Object {
            "color": "#2D3B45",
            "fontSize": 16,
            "lineHeight": 19,
            "paddingBottom": 8,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 10,
          }
        }
        testID="compose-message.body-text-input"
        underlineColorAndroid="transparent"
      />
    </KeyboardAwareScrollView>
  </View>
</Screen>
`;

exports[`Compose renders reply 1`] = `
<Screen
  drawUnderNavBar={true}
  leftBarButtons={
    Array [
      Object {
        "action": [Function],
        "testID": "compose-message.cancel",
        "title": "Cancel",
      },
    ]
  }
  rightBarButtons={
    Array [
      Object {
        "action": [Function],
        "disabled": true,
        "style": "done",
        "testID": "compose-message.send",
        "title": "Send",
      },
      Object {
        "accessibilityLabel": "Edit attachments (0)",
        "action": [Function],
        "badge": false,
        "image": Object {
          "testUri": "../../../src/images/paperclip-icon.png",
        },
        "testID": "compose-message.attach",
      },
    ]
  }
  showDismissButton={false}
  title="New Message"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <ModalOverlay
      activityIndicatorColor="#fff"
      animationType="fade"
      backgroundColor="rgba(0,0,0,0.8)"
      height={170}
      showActivityIndicator={true}
      text="Sending..."
      transparent={true}
      visible={false}
      width={170}
    />
    <KeyboardAwareScrollView
      contentContainerStyle={
        Object {
          "flexGrow": 1,
          "paddingBottom": 16,
        }
      }
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <ForwardRef
        style={
          Object {
            "borderBottomColor": "#C7CDD1",
            "borderBottomWidth": 0.5,
            "paddingHorizontal": 16,
          }
        }
        testID="compose.course-select"
        underlayColor="#ffffff00"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 54,
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#8B969E",
                  "fontSize": 16,
                  "lineHeight": 19,
                },
                Object {
                  "color": "#2D3B45",
                },
              ]
            }
          >
            Course 1
          </Text>
        </View>
      </ForwardRef>
      <View
        style={
          Array [
            Object {
              "borderBottomColor": "#C7CDD1",
              "borderBottomWidth": 0.5,
              "paddingHorizontal": 16,
            },
            Object {
              "alignItems": "flex-start",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "minHeight": 54,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "height": 54,
              "justifyContent": "center",
              "padding": 6,
              "paddingLeft": 0,
            }
          }
          testID="compose.recipients-placeholder"
        >
          <Text
            style={
              Object {
                "color": "#8B969E",
                "fontSize": 16,
                "lineHeight": 19,
              }
            }
          >
            To
          </Text>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "flexWrap": "wrap",
              "paddingVertical": 8,
            }
          }
        />
        <TouchableOpacity
          accessibilityLabel="Add recipient"
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          onPress={[Function]}
          style={
            Object {
              "height": 54,
              "justifyContent": "center",
            }
          }
          testID="compose.add-recipient"
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/images/Add.png",
              }
            }
            style={
              Object {
                "tintColor": undefined,
              }
            }
          />
        </TouchableOpacity>
      </View>
      <View
        style={
          Object {
            "borderBottomColor": "#C7CDD1",
            "borderBottomWidth": 0.5,
            "paddingHorizontal": 16,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={false}
          onChangeText={[Function]}
          placeholder="Subject"
          placeholderTextColor="#8B969E"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "color": "#2D3B45",
                "fontSize": 16,
                "height": 54,
                "lineHeight": 19,
              },
              Object {
                "color": "#8B969E",
                "fontSize": 16,
                "lineHeight": 19,
              },
              Object {
                "color": "#2D3B45",
              },
            ]
          }
          testID="compose-message.subject-text-input"
          underlineColorAndroid="transparent"
          value="(no subject)"
        />
      </View>
      <RowWithSwitch
        border="bottom"
        identifier="compose-message.send-all-toggle"
        onValueChange={[Function]}
        title="Send individual message to each recipient"
        value={false}
      />
      <TextInput
        allowFontScaling={true}
        multiline={true}
        onChangeText={[Function]}
        placeholder="Compose Message"
        placeholderTextColor="#8B969E"
        rejectResponderTermination={true}
        scrollEnabled={false}
        style={
          Object {
            "color": "#2D3B45",
            "fontSize": 16,
            "lineHeight": 19,
            "paddingBottom": 8,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 10,
          }
        }
        testID="compose-message.body-text-input"
        underlineColorAndroid="transparent"
      />
    </KeyboardAwareScrollView>
  </View>
</Screen>
`;

exports[`Compose renders with passed in recipients 1`] = `
<Screen
  drawUnderNavBar={true}
  leftBarButtons={
    Array [
      Object {
        "action": [Function],
        "testID": "compose-message.cancel",
        "title": "Cancel",
      },
    ]
  }
  rightBarButtons={
    Array [
      Object {
        "action": [Function],
        "disabled": true,
        "style": "done",
        "testID": "compose-message.send",
        "title": "Send",
      },
      Object {
        "accessibilityLabel": "Edit attachments (0)",
        "action": [Function],
        "badge": false,
        "image": Object {
          "testUri": "../../../src/images/paperclip-icon.png",
        },
        "testID": "compose-message.attach",
      },
    ]
  }
  showDismissButton={false}
  title="New Message"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <ModalOverlay
      activityIndicatorColor="#fff"
      animationType="fade"
      backgroundColor="rgba(0,0,0,0.8)"
      height={170}
      showActivityIndicator={true}
      text="Sending..."
      transparent={true}
      visible={false}
      width={170}
    />
    <KeyboardAwareScrollView
      contentContainerStyle={
        Object {
          "flexGrow": 1,
          "paddingBottom": 16,
        }
      }
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <ForwardRef
        onPress={[Function]}
        style={
          Object {
            "borderBottomColor": "#C7CDD1",
            "borderBottomWidth": 0.5,
            "paddingHorizontal": 16,
          }
        }
        testID="compose.course-select"
        underlayColor="#ffffff00"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 54,
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#8B969E",
                  "fontSize": 16,
                  "lineHeight": 19,
                },
                undefined,
              ]
            }
          >
            Select a Course
          </Text>
          <DisclosureIndicator />
        </View>
      </ForwardRef>
      <View
        style={
          Object {
            "borderBottomColor": "#C7CDD1",
            "borderBottomWidth": 0.5,
            "paddingHorizontal": 16,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          onChangeText={[Function]}
          placeholder="Subject"
          placeholderTextColor="#8B969E"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "color": "#2D3B45",
                "fontSize": 16,
                "height": 54,
                "lineHeight": 19,
              },
              Object {
                "color": "#8B969E",
                "fontSize": 16,
                "lineHeight": 19,
              },
              Object {
                "color": "#2D3B45",
              },
            ]
          }
          testID="compose-message.subject-text-input"
          underlineColorAndroid="transparent"
          value={null}
        />
      </View>
      <RowWithSwitch
        border="bottom"
        identifier="compose-message.send-all-toggle"
        onValueChange={[Function]}
        title="Send individual message to each recipient"
        value={false}
      />
      <TextInput
        allowFontScaling={true}
        multiline={true}
        onChangeText={[Function]}
        placeholder="Compose Message"
        placeholderTextColor="#8B969E"
        rejectResponderTermination={true}
        scrollEnabled={false}
        style={
          Object {
            "color": "#2D3B45",
            "fontSize": 16,
            "lineHeight": 19,
            "paddingBottom": 8,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 10,
          }
        }
        testID="compose-message.body-text-input"
        underlineColorAndroid="transparent"
      />
    </KeyboardAwareScrollView>
  </View>
</Screen>
`;
